`SerializableTestResultStore` flaky crash in `OutputRanges` when running Kotlin/Wasm tests

### Current Behavior

When running Kotlin/Wasm (wasmWasi) tests, the `wasmWasiTest` task intermittently fails with an IllegalStateException in Gradle's new `SerializableTestResultStore` / `OutputRanges` infrastructure. The tests themselves pass, but Gradle fails to serialize the test results into the binary format (`results-generic.bin`), causing the entire task to fail.

This is a flaky failure. It does not reproduce on every run, but occurs frequently enough to regularly break CI builds. In my project I see it fail roughly every few CI runs on GitHub Actions. The same test suite passes on re-run without any code changes, confirming this is a non-deterministic issue in Gradle's test result serialization, not in the tests themselves.

Example failure here: https://github.com/ZacSweers/metro/actions/runs/22003944676/job/63583152306 and scan https://scans.gradle.com/s/ifc43cje6ncws

```

Error while processing test process output message "##teamcity[testFinished name='invokesProvidersEveryTime' flowId='wasmTcAdapter1839979790']
"
java.lang.IllegalStateException: Output range end 0 specified without any output range starts
	at org.gradle.api.internal.tasks.testing.results.serializable.OutputRanges.of(OutputRanges.java:100)
	at org.gradle.api.internal.tasks.testing.results.serializable.TestOutputWriter.finishOutput(TestOutputWriter.java:85)
	at org.gradle.api.internal.tasks.testing.results.serializable.SerializableTestResultStore$Writer.completed(SerializableTestResultStore.java:219)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base/java.lang.reflect.Method.invoke(Method.java:565)
	at org.gradle.internal.dispatch.MethodInvocation.invokeOn(MethodInvocation.java:77)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:28)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:19)
	at org.gradle.internal.event.AbstractBroadcastDispatch.dispatch(AbstractBroadcastDispatch.java:44)
	at org.gradle.internal.event.BroadcastDispatch$SingletonDispatch.dispatch(BroadcastDispatch.java:270)
	at org.gradle.internal.event.BroadcastDispatch$SingletonDispatch.dispatch(BroadcastDispatch.java:172)
	at org.gradle.internal.event.AbstractBroadcastDispatch.dispatch(AbstractBroadcastDispatch.java:84)
	at org.gradle.internal.event.AbstractBroadcastDispatch.dispatch(AbstractBroadcastDispatch.java:70)
	at org.gradle.internal.event.BroadcastDispatch$CompositeDispatch.dispatch(BroadcastDispatch.java:382)
	at org.gradle.internal.event.BroadcastDispatch$CompositeDispatch.dispatch(BroadcastDispatch.java:274)
	at org.gradle.internal.event.ListenerBroadcast.dispatch(ListenerBroadcast.java:160)
	at org.gradle.internal.event.ListenerBroadcast.dispatch(ListenerBroadcast.java:37)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:88)
	at jdk.proxy3/jdk.proxy3.$Proxy204.completed(Unknown Source)
	at org.gradle.api.internal.tasks.testing.DefaultTestEventReporter.succeeded(DefaultTestEventReporter.java:95)
	at org.gradle.api.internal.tasks.testing.LifecycleTrackingTestEventReporter.succeeded(LifecycleTrackingTestEventReporter.java:72)
	at org.gradle.api.internal.tasks.testing.junit.result.TestEventReporterAsListener.completed(TestEventReporterAsListener.java:84)
	at org.gradle.api.internal.tasks.testing.results.StateTrackingTestResultProcessor.completed(StateTrackingTestResultProcessor.java:98)
	at org.jetbrains.kotlin.gradle.internal.testing.TCServiceMessagesClient$Node.reportCompleted(TCServiceMessagesClient.kt:373)
	at org.jetbrains.kotlin.gradle.internal.testing.TCServiceMessagesClient$TestNode.markCompleted(TCServiceMessagesClient.kt:493)
	at org.jetbrains.kotlin.gradle.internal.testing.TCServiceMessagesClient.close(TCServiceMessagesClient.kt:251)
	at org.jetbrains.kotlin.gradle.internal.testing.TCServiceMessagesClient.endTest(TCServiceMessagesClient.kt:148)
	at org.jetbrains.kotlin.gradle.internal.testing.TCServiceMessagesClient.serviceMessage(TCServiceMessagesClient.kt:73)
	at jetbrains.buildServer.messages.serviceMessages.ServiceMessagesParser.parse(ServiceMessagesParser.java:54)
	at jetbrains.buildServer.messages.serviceMessages.ServiceMessage.parse(ServiceMessage.java:121)
	at org.jetbrains.kotlin.gradle.internal.testing.TCServiceMessageOutputStreamHandler.parse(TCServiceMessageOutputStreamHandler.kt:121)
	at org.jetbrains.kotlin.gradle.internal.testing.TCServiceMessageOutputStreamHandler.flushLine(TCServiceMessageOutputStreamHandler.kt:81)
	at org.jetbrains.kotlin.gradle.internal.testing.TCServiceMessageOutputStreamHandler.write(TCServiceMessageOutputStreamHandler.kt:61)
	at org.gradle.process.internal.streams.ExecOutputHandleRunner.writeBuffer(ExecOutputHandleRunner.java:97)
	at org.gradle.process.internal.streams.ExecOutputHandleRunner.forwardContent(ExecOutputHandleRunner.java:85)
	at org.gradle.process.internal.streams.ExecOutputHandleRunner.run(ExecOutputHandleRunner.java:64)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.AbstractManagedExecutor$1.run(AbstractManagedExecutor.java:47)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1095)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:619)
	at java.base/java.lang.Thread.run(Thread.java:1447)
java.lang.IllegalStateException: Range end 97 is before stderr start 15293
(Use `node --trace-warnings ...` to show where the warning was created)

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':runtime:wasmWasiTest'.
> Could not write test report for results in [/home/runner/work/metro/metro/runtime/build/test-results/wasmWasiNodeTest/binary]
```

### Expected Behavior

The `wasmWasiTest` task should complete successfully when all tests pass. Gradle's test result serialization should handle the output from Kotlin's `TCServiceMessagesClient` without crashing, even when a test produces no stdout/stderr output or when output ranges arrive in unexpected order.

### Context (optional)

This is a flaky failure that breaks CI builds for Kotlin Multiplatform projects with `wasmWasi` targets. The tests themselves pass, the failure is entirely in Gradle's result serialization layer. This makes the failure particularly disruptive because it's not actionable from the project side and there is no way to disable the new binary result format. The only current workaround is downgrading to Gradle 9.2.x.

The issue appears to be related to the new test event reporting infrastructure introduced in Gradle 9.3.0.

Two distinct violations are thrown:
1. `OutputRanges.of()` rejects the case where a range end is specified without any range starts (a test completed with no output).
2. A separate validation rejects the case where `Range end 97 is before stderr start 15293` (output ranges are out of order).

This suggests either a race condition in how output events are tracked, or that `OutputRanges` doesn't account for tests that complete without producing any output — which is a valid case for Kotlin/Wasm tests reporting via TeamCity service messages.

Possibly related: #36601 (other `results-generic.bin` issues in 9.3.x).

### Self-contained Reproducer Project

This can be reproduced using the https://github.com/ZacSweers/metro project

```
$ git clone https://github.com/ZacSweers/metro.git
$ cd metro
$ ./gradlew :runtime:wasmWasiTest
```

Because this is a flaky issue, it may not fail on the first run. To increase the likelihood of reproducing, run it in a loop:

```bash
for i in $(seq 1 20); do
  echo "=== Run $i ==="
  ./gradlew :runtime:wasmWasiTest --rerun-tasks 2>&1 | tail -5
  if [ $? -ne 0 ]; then
    echo "FAILED on run $i"
    break
  fi
done
```

### Gradle version

9.3.1

### Gradle version that used to work

9.2.1

### Build scan URL (optional)

https://scans.gradle.com/s/ifc43cje6ncws

### Your Environment (optional)

- OS: Ubuntu (GitHub Actions `ubuntu-latest`)
- JDK: 25
- Kotlin: multiple versions
- Kotlin target: `wasmWasi` (using nodejs())
- Test framework: `kotlin.test` (routed through Kotlin's `TCServiceMessagesClient`)

--- a/platforms/software/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/results/SynchronizedTestListenerInternal.java
+++ b/platforms/software/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/results/SynchronizedTestListenerInternal.java
@@ -1,0 +1,66 @@
+/*
+ * Copyright 2026 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.tasks.testing.results;
+
+import org.gradle.api.internal.tasks.testing.TestCompleteEvent;
+import org.gradle.api.internal.tasks.testing.TestDescriptorInternal;
+import org.gradle.api.internal.tasks.testing.TestStartEvent;
+import org.gradle.api.tasks.testing.TestMetadataEvent;
+import org.gradle.api.tasks.testing.TestOutputEvent;
+import org.gradle.api.tasks.testing.TestResult;
+import org.jspecify.annotations.NullMarked;
+
+/**
+ * Wrapper to synchronize over all calls to a delegate {@link TestListenerInternal}.
+ */
+@NullMarked
+public final class SynchronizedTestListenerInternal implements TestListenerInternal {
+    private final TestListenerInternal delegate;
+    private final Object lock = new Object();
+
+    public SynchronizedTestListenerInternal(TestListenerInternal delegate) {
+        this.delegate = delegate;
+    }
+
+    @Override
+    public void started(TestDescriptorInternal testDescriptor, TestStartEvent startEvent) {
+        synchronized (lock) {
+            delegate.started(testDescriptor, startEvent);
+        }
+    }
+
+    @Override
+    public void completed(TestDescriptorInternal testDescriptor, TestResult testResult, TestCompleteEvent completeEvent) {
+        synchronized (lock) {
+            delegate.completed(testDescriptor, testResult, completeEvent);
+        }
+    }
+
+    @Override
+    public void output(TestDescriptorInternal testDescriptor, TestOutputEvent event) {
+        synchronized (lock) {
+            delegate.output(testDescriptor, event);
+        }
+    }
+
+    @Override
+    public void metadata(TestDescriptorInternal testDescriptor, TestMetadataEvent event) {
+        synchronized (lock) {
+            delegate.metadata(testDescriptor, event);
+        }
+    }
+}

--- a/platforms/software/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java
+++ b/platforms/software/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java
@@ -49,6 +49,7 @@ import org.gradle.api.internal.tasks.testing.report.generic.GenericHtmlTestRepor
 import org.gradle.api.internal.tasks.testing.report.generic.JunitXmlTestReportGenerator;
 import org.gradle.api.internal.tasks.testing.report.generic.TestTreeModelResultsProvider;
 import org.gradle.api.internal.tasks.testing.results.StateTrackingTestResultProcessor;
+import org.gradle.api.internal.tasks.testing.results.SynchronizedTestListenerInternal;
 import org.gradle.api.internal.tasks.testing.results.TestListenerAdapter;
 import org.gradle.api.internal.tasks.testing.results.TestListenerInternal;
 import org.gradle.api.invocation.Gradle;
@@ -588,7 +589,11 @@ public abstract class AbstractTestTask extends ConventionTask implements Verific
             false
         ))) {
             TestExecuter<TestExecutionSpec> testExecuter = Cast.uncheckedNonnullCast(createTestExecuter());
-            TestListenerInternal resultProcessorDelegate = reporterAsListener;
+            // Note: We don't have concurrency that requires a synchronized listener, but plugins that replace the test executer may.
+            // KGP is known to be one such plugin, see their `TCServiceMessagesClient`.
+            // This wrapper does not guarantee correctness from other parts of the system, but it restores previous behavior.
+            // It is recommended that plugins handle synchronization themselves and migrate to the new TestEventReporter APIs.
+            TestListenerInternal resultProcessorDelegate = new SynchronizedTestListenerInternal(reporterAsListener);
             if (failFast) {
                 resultProcessorDelegate = new FailFastTestListenerInternal(testExecuter, resultProcessorDelegate);
             }

